builtin/providers/aws/resource_aws_lambda_function.go

```diff
@@ -287,7 +279,96 @@ func resourceAwsLambdaFunctionDelete(d *schema.ResourceData, meta interface{}) e
 // resourceAwsLambdaFunctionUpdate maps to:
 // UpdateFunctionCode in the API / SDK
 func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	conn := meta.(*AWSClient).lambdaconn
+
+	d.Partial(true)
+
+	codeReq := &lambda.UpdateFunctionCodeInput{
+		FunctionName: aws.String(d.Id()),
+	}
+
+	codeUpdate := false
+	if v, ok := d.GetOk("filename"); ok && d.HasChange("source_code_hash") {
+		file, err := loadFileContent(v.(string))
+		if err != nil {
+			return fmt.Errorf("Unable to load %q: %s", v.(string), err)
+		}
+		codeReq.ZipFile = file
+		codeUpdate = true
+	}
+	if d.HasChange("s3_bucket") || d.HasChange("s3_key") || d.HasChange("s3_object_version") {
+		d.SetPartial("s3_bucket")
+		d.SetPartial("s3_key")
+		d.SetPartial("s3_object_version")
```

Comment:
Ah sorry yes, missed that. So I think the pattern should then be to `setPartial` on all the attributes that would be successfully updated by `UpdateFunctionCode` once it completes successfully. In pseudocode:

```
enable partial mode
build up UpdateFunctionCode request
  if error, return and nothing is updated
  if successful, mark all FunctionCode related attributes successful
build up UpdateFunctionConfig request
  if error, return and only marked FunctionCode attrs will be picked up
  if successful, disable partial mode, since everything worked (perhaps mark specific attrs here anyways to prepare for a potential 3rd update API call down the road?)
```

Let me know if this makes sense!